### PR TITLE
CCE GH boundary component part 2

### PR DESCRIPTION
**Pull request by moxcodes**
## Proposed changes

Split off from #2132 , completes the set of utilities needed for minimal CCE-GH interface.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Dependencies
- [x] #2132

### CODE REVIEW

File: src/Evolution/Systems/Cce/Actions/BoundaryComputeAndSendToEvolution.hpp [hunk 1]
```diff
@@ -72,5 +108,87 @@ struct BoundaryComputeAndSendToEvolution {
         true);
   }
 };
+
+/*!
+ * \ingroup ActionsGroup
+ * \brief Submits a request for CCE boundary data at the specified `time` to the
+ * `Cce::GhWorldtubeInterfaceManager`, and sends to the `EvolutionComponent`
+ * (template argument) it is ready.
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/BoundaryComputeAndSendToEvolution.hpp]: when/if  it is ready?

File: src/Evolution/Systems/Cce/Actions/BoundaryComputeAndSendToEvolution.hpp [hunk 2]
```diff
@@ -72,5 +108,87 @@ struct BoundaryComputeAndSendToEvolution {
...
+ *
+ * \details This uses the `Cce::GhWorldtubeInterfaceManager` to perform all of
+ * the work of managing the buffer of data sent from the Gh system and
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/BoundaryComputeAndSendToEvolution.hpp]: GH system

File: src/Evolution/Systems/Cce/Actions/BoundaryComputeAndSendToEvolution.hpp [hunk 3]
```diff
@@ -72,5 +108,87 @@ struct BoundaryComputeAndSendToEvolution {
...
+ * interpolating if necessary and supported. This dispatches then to
+ * `Cce::Actions::SendToEvolution<GhWorldtubeBoundary<Metavariables>,
+ * EvolutionComponent>` if the boundary data is ready, otherwise
+ * simply submitting the request and waiting for data to become available via
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/BoundaryComputeAndSendToEvolution.hpp]: use same tensing of verbs in both clauses (e.g. submits ... and waits. to match dispatches)

File: src/Evolution/Systems/Cce/Actions/BoundaryComputeAndSendToEvolution.hpp [hunk 4]
```diff
@@ -72,5 +108,87 @@ struct BoundaryComputeAndSendToEvolution {
...
+                    typename Metavariables::cce_boundary_communication_tags>,
+                DbTags>...>> = nullptr>
+  static void apply(db::DataBox<tmpl::list<DbTags...>>& box,
+                    Parallel::ConstGlobalCache<Metavariables>& cache,
+                    const ArrayIndex& /*array_index*/,
+                    const TimeStepId& time) noexcept {
+    db::mutate<Tags::GhInterfaceManager>(
+        make_not_null(&box),
+        [&time, &
+         cache ](const gsl::not_null<db::item_type<Tags::GhInterfaceManager>*>
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/BoundaryComputeAndSendToEvolution.hpp]: can you avoid the Item_type?

File: src/Evolution/Systems/Cce/Actions/BoundaryComputeAndSendToEvolution.hpp [hunk 5]
```diff
@@ -72,5 +108,87 @@ struct BoundaryComputeAndSendToEvolution {
...
+                     interface_manager) noexcept {
+          (*interface_manager)->request_gh_data(time);
+          if (const auto gh_data =
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/BoundaryComputeAndSendToEvolution.hpp]: move the assignment out of the `if()`, and then `static_cast<bool>` the result in the `if()` to make it clear that the validity of the data is being tested

File: src/Evolution/Systems/Cce/Actions/BoundaryComputeAndSendToEvolution.hpp [hunk 6]
```diff
@@ -72,5 +108,87 @@ struct BoundaryComputeAndSendToEvolution {
...
+                DbTags>...>> = nullptr>
+  static void apply(db::DataBox<tmpl::list<DbTags...>>& box,
+                    Parallel::ConstGlobalCache<Metavariables>& cache,
+                    const ArrayIndex& /*array_index*/, const TimeStepId& time,
+                    const tnsr::aa<DataVector, 3>& spacetime_metric,
+                    const tnsr::iaa<DataVector, 3>& phi,
+                    const tnsr::aa<DataVector, 3>& pi) noexcept {
+    create_bondi_boundary_data(
+        make_not_null(&box), phi, pi, spacetime_metric,
+        Parallel::get<InitializationTags::ExtractionRadius>(cache),
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/BoundaryComputeAndSendToEvolution.hpp]: get these from box

File: src/Evolution/Systems/Cce/Actions/ReceiveGhWorldtubeData.hpp [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "DataStructures/Variables.hpp"
+#include "Evolution/Systems/Cce/Actions/ReceiveWorldtubeData.hpp"
+#include "Evolution/Systems/Cce/InterfaceManagers/WorldtubeInterfaceManager.hpp"
+#include "Evolution/Systems/Cce/ReadBoundaryDataH5.hpp"
+#include "Parallel/ConstGlobalCache.hpp"
+#include "Parallel/Exit.hpp"
+#include "Parallel/Invoke.hpp"
+#include "Parallel/Printf.hpp"
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/ReceiveGhWorldtubeData.hpp]: check the includes, you don't use Exit, Printf, you do use Tensor, DataVector, Requires

File: src/Evolution/Systems/Cce/Actions/ReceiveGhWorldtubeData.hpp [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+      const tnsr::aa<DataVector, 3>& dt_spacetime_metric =
+          tnsr::aa<DataVector, 3>{},
+      const tnsr::iaa<DataVector, 3>& dt_phi = tnsr::iaa<DataVector, 3>{},
+      const tnsr::aa<DataVector, 3>& dt_pi =
+          tnsr::aa<DataVector, 3>{}) noexcept {
+    db::mutate<Tags::GhInterfaceManager>(
+        make_not_null(&box),
+        [&spacetime_metric, &phi, &pi, &dt_spacetime_metric, &dt_phi, &dt_pi,
+         &time,
+         &cache](const gsl::not_null<db::item_type<Tags::GhInterfaceManager>*>
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/ReceiveGhWorldtubeData.hpp]: can you avoid item_type?

File: src/Evolution/Systems/Cce/Actions/ReceiveGhWorldtubeData.hpp [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+                     interface_manager) noexcept {
+          (*interface_manager)
+              ->insert_gh_data(time, spacetime_metric, phi, pi,
+                               dt_spacetime_metric, dt_phi, dt_pi);
+          if (const auto gh_data =
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/ReceiveGhWorldtubeData.hpp]: same comment about the `if()` as above

File: tests/Unit/Evolution/Systems/Cce/Actions/Test_GhBoundaryCommunication.cpp [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+  using metavariables = Metavariables;
+  using chare_type = ActionTesting::MockArrayChare;
+  using array_index = size_t;
+
+  using simple_tags = tmpl::list<>;
+  using phase_dependent_action_list = tmpl::list<
+      Parallel::PhaseActions<typename Metavariables::Phase,
+                             Metavariables::Phase::Initialization,
+                             initialize_action_list>,
+      Parallel::PhaseActions<typename Metavariables::Phase,
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/Cce/Actions/Test_GhBoundaryCommunication.cpp]: don't need to list phases with empty action lists; only requirement is that one phase appears in the `phase_dependent_action_list`

**Review by kidder — CHANGES_REQUESTED** [review]
reviewed with @nilsdeppe

**Comment by moxcodes** [comment]
Fixups posted (and rebased on develop).
Cheers!

**Review by kidder — COMMENTED** [review]
squash

**Comment by moxcodes** [comment]
Squashed.
Cheers!

**Review by nilsdeppe — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**